File: apps/server/src/mentor_variants.ts
import type { Cost, MentorVariant } from "@divineruin/shared";
import { sql } from "./db.ts";
import { asRecord } from "./parse-helpers.ts";

// DB-loaded mentor variant catalog (M9 / story-001). Mirrors spells.ts/abilities.ts:
// content/mentor_variants.json -> mentor_variants table, loaded at startup, parsed
// fail-loud, exposed via getMentorVariant/getVariant/getVariantsForAbility. The Python
// agent reads the same table (apps/agent/mentor_variants.py); the mentor_variants.json
// row IS the cross-language contract and each loader owns fail-loud validation. This
// parser stays structurally symmetric with Python parse_mentor_variant_row.
//
// A variant is FULLY specified (decision m9 override shape): its cost/effect/narration
// replace the base ability's wholesale on activation (story-003). cost reuses the shared
// Cost {stamina, focus, scaling} object from the abilities layer.

// Runtime-loaded variants, keyed by variant id (populated by loadMentorVariants at startup).
let mentorVariants: ReadonlyMap<string, MentorVariant> = new Map();

// getMentorVariant is fail-loud (raises on unknown); getVariantsForAbility is a list
// query returning [] for unknown — mirroring Python get_mentor_variant /
// get_variants_for_ability.
export function getMentorVariant(id: string): MentorVariant {
  const variant = mentorVariants.get(id);
  if (!variant) throw new Error(`Unknown mentor variant: ${id}`);
  return variant;
}

// The contracted accessor for the activation path (story-003): validates the variant
// exists AND belongs to the named base ability, so a stored active variant id that
// doesn't match the ability fails loud rather than overriding the wrong technique.
export function getVariant(abilityId: string, variantId: string): MentorVariant {
  const variant = getMentorVariant(variantId);
  if (variant.ability_id !== abilityId) {
    throw new Error(
      `Mentor variant ${variantId} belongs to ${variant.ability_id}, not ${abilityId}`,
    );
  }
  return variant;
}

export function getVariantsForAbility(abilityId: string): MentorVariant[] {
  return Array.from(mentorVariants.values()).filter((v) => v.ability_id === abilityId);
}

export function setMentorVariants(map: ReadonlyMap<string, MentorVariant>): void {
  mentorVariants = map;
}

function parseCost(raw: unknown, ctx: string): Cost {
  const c = asRecord(raw, ctx);
  // Integer (not just number) for parity with the Python loader's _parse_cost, which
  // requires int — a shared row with a float cost must fail identically on both sides.
  if (typeof c.stamina !== "number" || !Number.isInteger(c.stamina)) {
    throw new Error(`${ctx}.stamina is not an integer`);
  }
  if (typeof c.focus !== "number" || !Number.isInteger(c.focus)) {
    throw new Error(`${ctx}.focus is not an integer`);
  }
  if (c.scaling !== null && typeof c.scaling !== "string") {
    throw new Error(`${ctx}.scaling is not a string or null`);
  }
  return { stamina: c.stamina, focus: c.focus, scaling: c.scaling };
}

export function parseMentorVariantRow(id: string, raw: unknown): MentorVariant {
  const data = asRecord(raw, `mentor_variants[${id}]`);
  const ctx = `mentor_variants[${id}]`;

  if (typeof data.ability_id !== "string") throw new Error(`${ctx}.ability_id is not a string`);
  if (typeof data.mentor_id !== "string") throw new Error(`${ctx}.mentor_id is not a string`);
  if (typeof data.effect !== "string") throw new Error(`${ctx}.effect is not a string`);
  if (typeof data.narration_cue !== "string")
    throw new Error(`${ctx}.narration_cue is not a string`);
  if (typeof data.cultural_attribution !== "string") {
    throw new Error(`${ctx}.cultural_attribution is not a string`);
  }

  return {
    id,
    ability_id: data.ability_id,
    mentor_id: data.mentor_id,
    cost: parseCost(data.cost, `${ctx}.cost`),
    effect: data.effect,
    narration_cue: data.narration_cue,
    cultural_attribution: data.cultural_attribution,
  };
}

export async function loadMentorVariants(): Promise<void> {
  const rows = await sql<{ id: string; data: unknown }[]>`
    SELECT id, data FROM mentor_variants
  `;
  // Build the local map then swap in one synchronous step, so a malformed row fails
  // loud WITHOUT wiping an already-loaded map (mirrors the Python loader's discipline).
  const map = new Map<string, MentorVariant>();
  for (const row of rows) {
    map.set(row.id, parseMentorVariantRow(row.id, row.data));
  }
  mentorVariants = map;
  console.log(`Loaded ${map.size} mentor variants`);
}

File: apps/agent/mentor_variants.py
"""Mentor variants — DB-loaded content config (M9 / story-001).

content/mentor_variants.json is the single source of truth for martial style
VARIANTS: a mentor-taught variant of a base elective technique that fully
overrides the technique's cost/effect/narration on activation and carries a
cultural-attribution string for the DM voice. This module is the Python loader,
an exact mirror of spells.py/abilities.py: a module-global dict populated by
load_mentor_variants() at process startup (or set_mentor_variants() in tests), a
fail-loud parse_mentor_variant_row shared by the DB loader and the JSON test
fixture, and sync accessors.

A variant is FULLY specified (decision m9 override shape): its cost is the same
Cost{stamina, focus, scaling} object the abilities layer uses (reused from
abilities, not redefined), so the activation path (story-003) can swap a
variant's values in wholesale. Each variant keys to a base ability_id (a martial
elective) and a teaching mentor NPC. Downstream: training unlock (story-002),
activation override + cultural narration (story-003).
"""

import json
import logging
from dataclasses import dataclass

from abilities import Cost
from catalog_parse import parse_str

logger = logging.getLogger("divineruin.mentor_variants")


@dataclass(frozen=True)
class MentorVariant:
    id: str
    ability_id: str
    mentor_id: str
    cost: Cost
    effect: str
    narration_cue: str
    cultural_attribution: str


# Module-level runtime-loaded variants, keyed by variant id. Populated by
# load_mentor_variants() at startup, or by set_mentor_variants() in tests.
_mentor_variants: dict[str, MentorVariant] = {}


def _parse_cost(raw: object, ctx: str) -> Cost:
    """Parse the variant cost into the shared Cost shape, fail-loud.

    Mirrors abilities._parse_cost — each loader owns its own fail-loud validation
    for cross-language parity (the same discipline the abilities/spells loaders
    follow), so this small parse is intentional symmetry, not duplication.
    """
    if not isinstance(raw, dict):
        raise ValueError(f"{ctx} is not an object")
    stamina = raw["stamina"]
    focus = raw["focus"]
    scaling = raw["scaling"]
    # bool is a subclass of int — exclude it explicitly, parity with the TS guard.
    if not isinstance(stamina, int) or isinstance(stamina, bool):
        raise ValueError(f"{ctx}.stamina is not an int")
    if not isinstance(focus, int) or isinstance(focus, bool):
        raise ValueError(f"{ctx}.focus is not an int")
    if scaling is not None and not isinstance(scaling, str):
        raise ValueError(f"{ctx}.scaling is not a string or null")
    return Cost(stamina=stamina, focus=focus, scaling=scaling)


def parse_mentor_variant_row(variant_id: str, data: dict) -> MentorVariant:
    """Parse a raw dict (from JSON file or DB JSONB) into a MentorVariant.

    Shared by load_mentor_variants (DB) and the JSON test fixture. Raises
    ValueError wrapping the underlying error with the row id for context; owns
    fail-loud validation of the cost shape and the required string fields.
    """
    try:
        return MentorVariant(
            id=variant_id,
            ability_id=parse_str(data["ability_id"], f"{variant_id}.ability_id"),
            mentor_id=parse_str(data["mentor_id"], f"{variant_id}.mentor_id"),
            cost=_parse_cost(data["cost"], f"{variant_id}.cost"),
            effect=parse_str(data["effect"], f"{variant_id}.effect"),
            narration_cue=parse_str(data["narration_cue"], f"{variant_id}.narration_cue"),
            cultural_attribution=parse_str(data["cultural_attribution"], f"{variant_id}.cultural_attribution"),
        )
    except (KeyError, TypeError) as e:
        raise ValueError(f"Malformed mentor_variants row {variant_id!r}: {e}") from e


def set_mentor_variants(config: dict[str, MentorVariant]) -> None:
    """Test seam: populate _mentor_variants directly without going through the DB."""
    _mentor_variants.clear()
    _mentor_variants.update(config)


def get_mentor_variant(variant_id: str) -> MentorVariant:
    """Return the variant for an id. Raises ValueError if not loaded/unknown."""
    if variant_id not in _mentor_variants:
        raise ValueError(f"Unknown mentor variant: {variant_id!r}")
    return _mentor_variants[variant_id]


def get_variant(ability_id: str, variant_id: str) -> MentorVariant:
    """Return the variant identified by (ability_id, variant_id), fail-loud.

    The contracted accessor for the activation path (story-003): it validates the
    variant exists AND belongs to the named base ability, so a stored active
    variant id that doesn't match the ability fails loud rather than overriding
    the wrong technique.
    """
    variant = get_mentor_variant(variant_id)
    if variant.ability_id != ability_id:
        raise ValueError(f"Mentor variant {variant_id!r} belongs to {variant.ability_id!r}, not {ability_id!r}")
    return variant


def get_variants_for_ability(ability_id: str) -> tuple[MentorVariant, ...]:
    """Return all loaded variants for a base ability, in load order.

    Empty tuple when the ability has none loaded (e.g. an unknown id), the
    ability-keyed analogue of spells.get_spells_by_source.
    """
    return tuple(v for v in _mentor_variants.values() if v.ability_id == ability_id)


def is_loaded() -> bool:
    """True once the variants have been populated (startup load or test seam)."""
    return bool(_mentor_variants)


async def load_mentor_variants() -> None:
    """Load the mentor variant catalog from the DB into _mentor_variants.

    Called at agent + async-worker startup beside load_spells(). Fails loud if the
    query or a row errors. Builds a local dict then swaps in one synchronous step
    (no await between), so a malformed row fails loud WITHOUT wiping an
    already-loaded map and a concurrent get_variant never observes a
    half-populated map.
    """
    import db

    pool = await db.get_pool()
    rows = await pool.fetch("SELECT id, data FROM mentor_variants")
    loaded: dict[str, MentorVariant] = {}
    for row in rows:
        data = json.loads(row["data"]) if isinstance(row["data"], str) else row["data"]
        loaded[row["id"]] = parse_mentor_variant_row(row["id"], data)
    _mentor_variants.clear()
    _mentor_variants.update(loaded)
    logger.info("Loaded %d mentor variants", len(_mentor_variants))
